2022/Q2/artifacts/pos-model/Staking.tla

```diff
@@ -113,60 +120,78 @@ Unbond(sender, amount) ==
     LET fail ==
         \/ amount < 0
         \/ sender = Validator
-        \/ amount > delegated[sender]
+        \/ amount > delegated[PipelineLength,sender]
     IN
     /\ lastTx' = [ id |-> nextTxId, tag |-> "unbond", fail |-> fail,
                    sender |-> sender, toAddr |-> Validator, value |-> amount ]
     /\ nextTxId' = nextTxId + 1
+    /\ txCounter' = txCounter + 1
     /\ failed' = (fail \/ failed)
+    /\ UNCHANGED epoch
     /\  IF fail
         THEN
           UNCHANGED <<balanceOf, unbonded, delegated>>
         ELSE
           \* transaction succeeds
-          /\ unbonded' = [ unbonded EXCEPT ![sender] = @ + amount ]
-          /\ delegated' = [ delegated EXCEPT ![sender] = @ - amount ]
+          /\ unbonded' = [ unbonded EXCEPT ![UnbondingLength, sender] = @ + amount ]
+          /\ delegated' = [ delegated EXCEPT ![PipelineLength, sender] = @ - amount ]
           /\ UNCHANGED  balanceOf
 
 \* withdraw unbonded coins
 Withdraw(sender) ==
     LET fail ==
         \/ sender = Validator
-        \/ unbonded[sender] <= 0
+        \/ unbonded[1, sender] <= 0
     IN
     /\ lastTx' = [ id |-> nextTxId, tag |-> "withdraw", fail |-> fail,
-                   sender |-> sender, toAddr |-> Validator, value |-> unbonded[sender] ]
+                   sender |-> sender, toAddr |-> Validator, value |-> unbonded[1, sender] ]
     /\ nextTxId' = nextTxId + 1
+    /\ txCounter' = txCounter + 1
```

Comment:
I see nothing wrong with doing what you suggest and it is probably neater. It is as it is because Igor's spec was doing it that way and I started from there ;) I have added a new operator Common that includes the actions common to all transactions